### PR TITLE
Improve client SNI in testSniWithApnHandler

### PR DESCRIPTION
**Pull request by sunrenjie**
Motivations
The test SniHandlerTest#testSniWithApnHandler() does not actually
involve SNI: given the client setup, the ClientHello in the form of hex
strings is not actually written to the wire, so the server never receives that.
We may need to write in somewhere else (e.g., channelActive()) instead of in
initChannel() in order for the hex strings to reach the server. So here
what's actually going on is an ordinary TLS C/S communication without SNI.

Modifications
The client part is modified to enable SNI by using an SslHandler with an
SSLEngine created by io.netty.handler.ssl.SslContext#newEngine(), where
the server hostname is specified. Also, more clauses are added to verify that
the SNI is indeed successful.

Results
Now the test verifies that both SNI and APN actually happen and succeed.

### CODE REVIEW

**Comment by sunrenjie** [comment]
@normanmaurer @Scottmitch PTAL

File: handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java [hunk 1]
```diff
@@ -212,11 +201,8 @@ protected void configurePipeline(ChannelHandlerContext ctx, String protocol) thr
             cb.handler(new ChannelInitializer<Channel>() {
                 @Override
                 protected void initChannel(Channel ch) throws Exception {
-                    // Simulate client side SNI (which is currently not supported).
-                    ch.write(Unpooled.wrappedBuffer(DatatypeConverter.parseHexBinary(tlsHandshakeMessageHex1)));
-                    ch.writeAndFlush(Unpooled.wrappedBuffer(DatatypeConverter.parseHexBinary(tlsHandshakeMessageHex)));
-                    // Add a SslHandler so we can do the client side of the handshake.
-                    ch.pipeline().addLast(clientContext.newHandler(ch.alloc()));
+                    ch.pipeline().addLast(new SslHandler(clientContext.newEngine(
+                            ByteBufAllocator.DEFAULT, "sni.fake.site", -1)));
```
Comment (Scottmitch) [inline handler/src/test/java/io/netty/handler/ssl/SniHandlerTest.java]: `ByteBufAllocator.DEFAULT` -> `ch.alloc()`

**Comment by Scottmitch** [comment]
@sunrenjie - thanks for the better unit test! 1 minor comment and lgtm

**Comment by sunrenjie** [comment]
@Scottmitch Done with your comment, thanks!
Actually, I shamelessly copied the code from here: https://github.com/netty/netty/blob/4.1/handler/src/test/java/io/netty/handler/ssl/SniClientTest.java#L91 . I guess that line needs change too ?

By the way, is there any tricky point in enabling client SNI via io.netty.handler.ssl.SslContext#newEngine(...) ? I was very reluctant to do so. :)

**Comment by normanmaurer** [comment]
Ok cool let me merge

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (6b427aaee5ca6409cf3783f0cdb26e0d680a871a) and 4.0 (ad273876467a0afb22ba34d7ea47f1f391446826)

**Pull request closed without merging**